aten/src/ATen/native/cpu/group_norm_kernel.cpp

```diff
@@ -785,6 +785,531 @@ void GroupNormBackwardKernelImplInternal(
+      int64_t n{0}, g{0};
+      data_index_init(begin, n, N, g, G);
+      for (const auto i : c10::irange(begin, end)) {
+        // Step 1. Compute internal gradients.
+        PT* ds_ptr = ds_data + i * D;
+        PT* db_ptr = db_data + i * D;
+        PT ds_gamma, db_gamma;
+        const T* X_ptr = X_data + n * HxW * C + g * D;
+        const T* dY_ptr = dY_data + n * HxW * C + g * D;
+        const PT* gamma_ptr = gamma_null ? gamma_data : (gamma_data + g * D);
+        std::tie(ds_gamma, db_gamma) = CalcInternalGradientsChannelsLast<T, PT>(
+          X_ptr,
+          dY_ptr,
+          gamma_ptr,
+          ds_ptr,
+          db_ptr,
+          HxW,
+          C,
+          D);
+
+        // Step 2. Compute dX.
+        T* dX_ptr = dX_data + n * HxW * C + g * D;
+        const PT* rstd_ptr = rstd_data + i;
+        const PT c2 =
+            (db_gamma * mean_data[i] - ds_gamma) * rstd_data[i] * rstd_data[i] * rstd_data[i] * s;
+        const PT c3 = -c2 * mean_data[i] - db_gamma * rstd_data[i] * s;
+        ApplyInputGradientsChannelsLast<T, PT>(dY_ptr, X_ptr, dX_ptr, rstd_ptr, gamma_ptr, c2, c3, HxW, C, D, true);
+        data_index_step(n, N, g, G);
+      }
+    });
+
+  } else {
+    // impl-2: parallel on N * HxW.
+    int num_threads = at::get_num_threads();
+    Tensor buffer = at::empty({num_threads, N, 2 * C},
+      X.options().dtype(c10::CppTypeToScalarType<PT>::value)).zero_();
+    PT* buffer_data = buffer.data_ptr<PT>();
+
+    Tensor tmp_buffer = at::empty({N, 2 * G},
+      X.options().dtype(c10::CppTypeToScalarType<PT>::value));
+    PT* tmp_buffer_data = tmp_buffer.data_ptr<PT>();
+
+    // Step 1. Each thread compute their own internal gradients to the buffer.
+    at::parallel_for(0, N * HxW, 1, [&](int64_t begin, int64_t end) {
+      int tid = at::get_thread_num();
+      PT* buffer_ptr = buffer_data + tid * N * 2 * C;
+      int64_t n{0}, m{0};
+      data_index_init(begin, n, N, m, HxW);
+      for (const auto i : c10::irange(begin, end)) {
+        PT* ds_ptr = buffer_ptr +  n * 2 * C;
+        PT* db_ptr = ds_ptr + C;
+        const T* X_ptr = X_data + i * C;
+        const T* dY_ptr = dY_data + i * C;
+
+        DsDbRowwiseMomentsChannelsLast<T, PT>(dY_ptr, X_ptr, ds_ptr, db_ptr, C);
+        data_index_step(n, N, m, HxW);
+      }
+    });
+
+    // Step 2. Collect internal gradients from each thread and
+    // get the final internal gradients to ds, db, and tmp_buffer.
+    for (const auto n : c10::irange(N)) {
+      for (const auto g : c10::irange(G)) {
+        PT ds_gamma{0}, db_gamma{0};
+        for (const auto d : c10::irange(D)) {
+          PT ds_val{0}, db_val{0};
+          for (const auto t : c10::irange(num_threads)) {
+            PT* buffer_ptr = buffer_data + t * N * 2 * C + n * 2 * C;
+            ds_gamma += buffer_ptr[g * D + d] * (gamma_null ? PT(1) : gamma_data[g * D + d]);
+            db_gamma += buffer_ptr[g * D + d + C] * (gamma_null ? PT(1) : gamma_data[g * D + d]);
```

Comment:
Nit (compiler should be smart enough to optimize those out, but makes code a bit more readable.)
```suggestion
             if (gamma_null) {
                         ds_gamma += buffer_ptr[g * D + d];
                         db_gamma += buffer_ptr[g * D + d + C];
             } else {
                ds_gamma += buffer_ptr[g * D + d] *  gamma_data[g * D + d];
                db_gamma += buffer_ptr[g * D + d + C] * gamma_data[g * D + d];             
             }
```

Reply:
Fixed it.